ext/json/json_encoder.c

```diff
@@ -104,7 +104,11 @@ static inline void php_json_encode_double(smart_str *buf, double d, int options)
 {
 	size_t len;
 	char num[PHP_JSON_DOUBLE_MAX_LENGTH];
-	php_gcvt(d, (int)EG(precision), '.', 'e', &num[0]);
+	if (JSON_G(precision) == -1) {
+		php_0cvt(d, 17, '.', 'e', num);
+	} else {
+		php_gcvt(d, JSON_G(precision), '.', 'e', num);
+	}
```

Comment:
Instead of checking this everywhere, maybe just support ndigits=-1 in php_gcvt? This should allow us to directly use this mode in printf etc by passing -1 (or EG(serialize_precision)).